### PR TITLE
SNC: Scheming Fence and support

### PR DESCRIPTION
**Pull request by Northmoc**
Human and AI mana conversion seems to work

I think AI needs some help with the ChooseCard line
In limited testing without `AILogic` the AI would sometimes choose one of its own cards with no activated abilities, so for now I did `AILogic$ OppPreferred`
This is a card that could have various strategic uses. These all seem like good picks:
1. lock down an opponent planeswalker (gaining no abilities)
2. lock down an opponent non-PW (gaining its activated abilities)
3. choose a difficult to activate permanent of your own (this would fix the color requirements)
So not sure how much of that is feasible to wire up for the AI – 3. in particular seems pretty obscure

### CODE REVIEW

File: forge-gui/res/cardsfolder/upcoming/scheming_fence.txt [hunk 1]
```diff
@@ -0,0 +1,9 @@
+Name:Scheming Fence
+ManaCost:W U
+Types:Creature Human Citizen
+PT:2/3
+K:ETBReplacement:Other:ChooseCard
+SVar:ChooseCard:DB$ ChooseCard | Choices$ Permanent.nonLand | MinAmount$ 0 | ChoiceTitle$ You may choose a nonland permanent | AILogic$ OppPreferred | SpellDescription$ As CARDNAME enters the battlefield, you may choose a nonland permanent.
+S:Mode$ Continuous | Affected$ Card.ChosenCard | AddHiddenKeyword$ CARDNAME's activated abilities can't be activated. | Description$ Activated abilities of the chosen permanent can't be activated.
+S:Mode$ Continuous | Affected$ Card.Self | EffectZone$ Battlefield | GainsAbilitiesOfDefined$ ChosenCard | GainsAbilitiesActivateIgnoreColor$ True | Description$ CARDNAME has all activated abilities of the chosen permanent except for loyalty abilities. You may spend mana as though it were mana of any color to activate those abilities.
```
Comment (Hanmac) [inline forge-gui/res/cardsfolder/upcoming/scheming_fence.txt:8]: `Except Loyatly Abilities`, you probably need a isValid filter in the Class

Comment (Northmoc) [reply forge-gui/res/cardsfolder/upcoming/scheming_fence.txt:8]: I thought this boolean would take care of that, but apparently not :(
image

Comment (Hanmac) [reply forge-gui/res/cardsfolder/upcoming/scheming_fence.txt:8]: yeah, better replace it with a Valid Check that can be better extended

File: forge-game/src/main/java/forge/game/staticability/StaticAbilityContinuous.java [hunk 2]
```diff
@@ -832,14 +831,20 @@ public String apply(String input) {
                     for (Card c : cardsIGainedAbilitiesFrom) {
                         for (SpellAbility sa : c.getSpellAbilities()) {
                             if (sa.isActivatedAbility()) {
-                                if (loyaltyAB && !sa.isPwAbility()) {
+                                if (params.containsKey("GainsValidAbilities") &&
+                                        !sa.isValid(params.get("GainsValidAbilities").split(","),
```
Comment (Hanmac) [inline forge-game/src/main/java/forge/game/staticability/StaticAbilityContinuous.java:835]: stAb.matchValidParam

**Comment by Hanmac** [comment]
What happens if opponent card gets bounced? does it still block the abilities or does it work again?

**Comment by Northmoc** [comment]
> What happens if opponent card gets bounced? does it still block the abilities or does it work again?

The abilities are still blocked – should we set a static to lose the ChosenCard if it changes zone?

**Comment by Hanmac** [comment]
> > What happens if opponent card gets bounced? does it still block the abilities or does it work again?
> 
> The abilities are still blocked – should we set a static to clean up the ChosenCard if it changes zone?

yeah, a cleanup trigger for now would be good, until i refactor it again, probably with Linked Ability stuff

**Review by tool4ever — APPROVED** [review]

**Pull request merged**